From eb726848cd4e096a815d53f86784f0e05cf70ed6 Mon Sep 17 00:00:00 2001
From: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Date: Mon, 17 Mar 2025 11:50:59 +0100
Subject: [PATCH] pwm: stm32-lp: rework to follow stm32mp25 upstream

Rework to follow changes done upstream on stm32mp25:
- rely on fallback compatible as no specific .data is associated to the
  driver. Matching is achieved by using MFD parent data.
- renamed registers/bits definitions

Signed-off-by: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Change-Id: I7f2f71210eb38abc3e9b2aafb4e0cfc9a244f5f7
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/445393
Reviewed-by: Olivier MOYSAN <olivier.moysan@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 drivers/pwm/pwm-stm32-lp.c | 56 ++++++++++++++++++--------------------
 1 file changed, 27 insertions(+), 29 deletions(-)

diff --git a/drivers/pwm/pwm-stm32-lp.c b/drivers/pwm/pwm-stm32-lp.c
index 7de64a41fea3..884911d443a6 100644
--- a/drivers/pwm/pwm-stm32-lp.c
+++ b/drivers/pwm/pwm-stm32-lp.c
@@ -43,10 +43,10 @@ static int stm32_pwm_lp_update_allowed(struct stm32_pwm_lp *priv, int channel)
 	if (!priv->num_cc_chans)
 		return true;
 
-	ret = regmap_read(priv->regmap, STM32MP25_LPTIM_CCMR1, &ccmr1);
+	ret = regmap_read(priv->regmap, STM32_LPTIM_CCMR1, &ccmr1);
 	if (ret)
 		return ret;
-	ccmr = ccmr1 & (STM32MP25_LPTIM_CC1E | STM32MP25_LPTIM_CC2E);
+	ccmr = ccmr1 & (STM32_LPTIM_CC1E | STM32_LPTIM_CC2E);
 
 	/* More than one channel enabled: enable, prescaler or ARR value can't be changed */
 	if (bitmap_weight(&ccmr, sizeof(u32) * BITS_PER_BYTE) > 1)
@@ -57,9 +57,9 @@ static int stm32_pwm_lp_update_allowed(struct stm32_pwm_lp *priv, int channel)
 	 * report if enable, prescaler or ARR value can be changed.
 	 */
 	if (channel)
-		return !(ccmr1 & STM32MP25_LPTIM_CC1E);
+		return !(ccmr1 & STM32_LPTIM_CC1E);
 	else
-		return !(ccmr1 & STM32MP25_LPTIM_CC2E);
+		return !(ccmr1 & STM32_LPTIM_CC2E);
 }
 
 static int stm32_pwm_lp_compare_channel_apply(struct stm32_pwm_lp *priv, int channel,
@@ -73,25 +73,25 @@ static int stm32_pwm_lp_compare_channel_apply(struct stm32_pwm_lp *priv, int cha
 	if (!priv->num_cc_chans)
 		return 0;
 
-	ret = regmap_read(priv->regmap, STM32MP25_LPTIM_CCMR1, &ccmr1);
+	ret = regmap_read(priv->regmap, STM32_LPTIM_CCMR1, &ccmr1);
 	if (ret)
 		return ret;
 
 	if (channel) {
-		/* Must disable CC channel (CCxE) to modify polarity (CCxP), then reenable */
-		reenable = (enable && FIELD_GET(STM32MP25_LPTIM_CC2E, ccmr1)) &&
-			(polarity != FIELD_GET(STM32MP25_LPTIM_CC2P, ccmr1));
+		/* Must disable CC channel (CCxE) to modify polarity (CCxP), then re-enable */
+		reenable = (enable && FIELD_GET(STM32_LPTIM_CC2E, ccmr1)) &&
+			(polarity != FIELD_GET(STM32_LPTIM_CC2P, ccmr1));
 
-		mask = STM32MP25_LPTIM_CC2SEL | STM32MP25_LPTIM_CC2E | STM32MP25_LPTIM_CC2P;
-		val = FIELD_PREP(STM32MP25_LPTIM_CC2P, polarity);
-		val |= FIELD_PREP(STM32MP25_LPTIM_CC2E, enable);
+		mask = STM32_LPTIM_CC2SEL | STM32_LPTIM_CC2E | STM32_LPTIM_CC2P;
+		val = FIELD_PREP(STM32_LPTIM_CC2P, polarity);
+		val |= FIELD_PREP(STM32_LPTIM_CC2E, enable);
 	} else {
-		reenable = (enable && FIELD_GET(STM32MP25_LPTIM_CC1E, ccmr1)) &&
-			(polarity != FIELD_GET(STM32MP25_LPTIM_CC1P, ccmr1));
+		reenable = (enable && FIELD_GET(STM32_LPTIM_CC1E, ccmr1)) &&
+			(polarity != FIELD_GET(STM32_LPTIM_CC1P, ccmr1));
 
-		mask = STM32MP25_LPTIM_CC1SEL | STM32MP25_LPTIM_CC1E | STM32MP25_LPTIM_CC1P;
-		val = FIELD_PREP(STM32MP25_LPTIM_CC1P, polarity);
-		val |= FIELD_PREP(STM32MP25_LPTIM_CC1E, enable);
+		mask = STM32_LPTIM_CC1SEL | STM32_LPTIM_CC1E | STM32_LPTIM_CC1P;
+		val = FIELD_PREP(STM32_LPTIM_CC1P, polarity);
+		val |= FIELD_PREP(STM32_LPTIM_CC1E, enable);
 	}
 
 	if (reenable) {
@@ -99,8 +99,8 @@ static int stm32_pwm_lp_compare_channel_apply(struct stm32_pwm_lp *priv, int cha
 		unsigned long rate;
 		unsigned int delay_us;
 
-		ret = regmap_update_bits(priv->regmap, STM32MP25_LPTIM_CCMR1,
-					 channel ? STM32MP25_LPTIM_CC2E : STM32MP25_LPTIM_CC1E, 0);
+		ret = regmap_update_bits(priv->regmap, STM32_LPTIM_CCMR1,
+					 channel ? STM32_LPTIM_CC2E : STM32_LPTIM_CC1E, 0);
 		if (ret)
 			return ret;
 		/*
@@ -118,7 +118,7 @@ static int stm32_pwm_lp_compare_channel_apply(struct stm32_pwm_lp *priv, int cha
 		usleep_range(delay_us, delay_us * 2);
 	}
 
-	return regmap_update_bits(priv->regmap, STM32MP25_LPTIM_CCMR1, mask, val);
+	return regmap_update_bits(priv->regmap, STM32_LPTIM_CCMR1, mask, val);
 }
 
 static int stm32_pwm_lp_apply(struct pwm_chip *chip, struct pwm_device *pwm,
@@ -142,7 +142,7 @@ static int stm32_pwm_lp_apply(struct pwm_chip *chip, struct pwm_device *pwm,
 			if (ret)
 				return ret;
 			ret = regmap_write(priv->regmap, pwm->hwpwm ?
-					   STM32MP25_LPTIM_CCR2 : STM32_LPTIM_CMP, 0);
+					   STM32_LPTIM_CCR2 : STM32_LPTIM_CMP, 0);
 			if (ret)
 				return ret;
 
@@ -252,7 +252,7 @@ static int stm32_pwm_lp_apply(struct pwm_chip *chip, struct pwm_device *pwm,
 		goto err;
 
 	/* Write CMP/CCRx register and ensure it's been properly written */
-	ret = regmap_write(priv->regmap, pwm->hwpwm ? STM32MP25_LPTIM_CCR2 : STM32_LPTIM_CMP,
+	ret = regmap_write(priv->regmap, pwm->hwpwm ? STM32_LPTIM_CCR2 : STM32_LPTIM_CMP,
 			   prd - (1 + dty));
 	if (ret)
 		goto err;
@@ -308,11 +308,11 @@ static int stm32_pwm_lp_get_state(struct pwm_chip *chip,
 	enabled = !!FIELD_GET(STM32_LPTIM_ENABLE, val);
 	if (priv->num_cc_chans) {
 		/* There's a CC chan, need to also check if it's enabled */
-		regmap_read(priv->regmap, STM32MP25_LPTIM_CCMR1, &ccmr1);
+		regmap_read(priv->regmap, STM32_LPTIM_CCMR1, &ccmr1);
 		if (pwm->hwpwm)
-			enabled &= !!FIELD_GET(STM32MP25_LPTIM_CC2E, ccmr1);
+			enabled &= !!FIELD_GET(STM32_LPTIM_CC2E, ccmr1);
 		else
-			enabled &= !!FIELD_GET(STM32MP25_LPTIM_CC1E, ccmr1);
+			enabled &= !!FIELD_GET(STM32_LPTIM_CC1E, ccmr1);
 	}
 	state->enabled = enabled;
 
@@ -327,9 +327,9 @@ static int stm32_pwm_lp_get_state(struct pwm_chip *chip,
 	presc = FIELD_GET(STM32_LPTIM_PRESC, val);
 	if (priv->num_cc_chans) {
 		if (pwm->hwpwm)
-			state->polarity = FIELD_GET(STM32MP25_LPTIM_CC2P, ccmr1);
+			state->polarity = FIELD_GET(STM32_LPTIM_CC2P, ccmr1);
 		else
-			state->polarity = FIELD_GET(STM32MP25_LPTIM_CC1P, ccmr1);
+			state->polarity = FIELD_GET(STM32_LPTIM_CC1P, ccmr1);
 	} else {
 		state->polarity = FIELD_GET(STM32_LPTIM_WAVPOL, val);
 	}
@@ -339,7 +339,7 @@ static int stm32_pwm_lp_get_state(struct pwm_chip *chip,
 	tmp = (tmp << presc) * NSEC_PER_SEC;
 	state->period = DIV_ROUND_CLOSEST_ULL(tmp, rate);
 
-	regmap_read(priv->regmap, pwm->hwpwm ? STM32MP25_LPTIM_CCR2 : STM32_LPTIM_CMP, &val);
+	regmap_read(priv->regmap, pwm->hwpwm ? STM32_LPTIM_CCR2 : STM32_LPTIM_CMP, &val);
 	tmp = prd - val;
 	tmp = (tmp << presc) * NSEC_PER_SEC;
 	state->duty_cycle = DIV_ROUND_CLOSEST_ULL(tmp, rate);
@@ -440,8 +440,6 @@ static const struct dev_pm_ops stm32_pwm_lp_pm_ops = {
 
 static const struct of_device_id stm32_pwm_lp_of_match[] = {
 	{ .compatible = "st,stm32-pwm-lp", },
-	{ .compatible = "st,stm32mp21-pwm-lp", },
-	{ .compatible = "st,stm32mp25-pwm-lp", },
 	{},
 };
 MODULE_DEVICE_TABLE(of, stm32_pwm_lp_of_match);
-- 
2.50.1

